File: lclang/src/main/java/lclang/LCCompiler.java
package lclang;

import org.antlr.v4.runtime.tree.ParseTree;
import postvm.Caller;
import postvm.Utils;
import postvm.exceptions.LCLangRuntimeException;
import postvm.library.classes.ArrayClass;
import postvm.library.classes.PostVMClass;
import postvm.library.classes.VoidClass;
import postvm.methods.MethodImpl;
import postvm.statements.*;
import postvm.statements.expressions.*;

import java.util.ArrayList;
import java.util.List;

public class LCCompiler extends lclangBaseVisitor<Statement> {
    public static final LCCompiler instance = new LCCompiler();
    private LCCompiler() {}

    @Override
    @Deprecated
    public Statement visit(ParseTree tree) {
        return super.visit(tree);
    }

    @Override
    public Statement visitStmt(lclangParser.StmtContext ctx) {
        return ctx.children.get(0).accept(this);
    }

    @Override
    public VariableExpression visitVariable(lclangParser.VariableContext ctx) {
        return VariableExpression.get(ctx.getText());
    }

    @Override
    public Statement visitBlock(lclangParser.BlockContext ctx) {
        Statement[] statements = new Statement[ctx.stmt().size()];
        for (int i = 0; i < statements.length; i++) {
            statements[i] = ctx.stmt().get(i).accept(this);
        }

        return new BlockStatement(statements);
    }

    @Override
    public Statement visitContainer(lclangParser.ContainerContext ctx) {
        Statement[] statements = new Statement[ctx.stmt().size()];
        for (int i = 0; i < statements.length; i++) {
            statements[i] = ctx.stmt().get(i).accept(this);
        }

        return new ContainerExpression(statements);
    }

    @Override
    public ValueExpression visitValue(lclangParser.ValueContext ctx) {

        String text = ctx.getText();
        if (ctx.STRING() != null)
            return ValueExpression.get(Utils.unescapeString(text.substring(1, text.length() - 1)));
        else if (ctx.CHAR() != null)
            return ValueExpression.get(Utils.unescapeString(text).charAt(1));
        else if (ctx.DOUBLE() != null)
            return ValueExpression.get(Double.parseDouble(text));
        else if (ctx.INTEGER() != null)
            return ValueExpression.get(Integer.parseInt(text));
        else if (ctx.LONG() != null)
            return ValueExpression.get(Long.parseLong(text));
        else if (ctx.HEX_LONG() != null)
            return ValueExpression.get(Long.parseLong(text, 16));
        else if (ctx.BOOL() != null)
            return ValueExpression.get(text.length() == 4);
        else throw new RuntimeException();
    }

    @Override
    public Statement visitWhileStmt(lclangParser.WhileStmtContext ctx) {
        Statement statement = null;
        lclangParser.StmtContext stmtCtx = ctx.stmt();
        if(stmtCtx!=null){
            statement = visitStmt(stmtCtx);
        }

        return new WhileStatement(visitExpression(ctx.expression()), statement);
    }

    @Override
    public Statement visitIfStmt(lclangParser.IfStmtContext ctx) {
        Statement ifFalse = null;
        if(ctx.ifF!=null)
            ifFalse = visitStmt(ctx.ifF);

        return new IfStatement(
                visitExpression(ctx.condition),
                visitStmt(ctx.ifT),
                ifFalse);
    }

    @Override
    public Statement visitIfExpr(lclangParser.IfExprContext ctx) {
        return new IfExpression(
                visitExpression(ctx.condition),
                visitExpression(ctx.ifT),
                visitExpression(ctx.ifF));
    }

    @Override
    public Statement visitArray(lclangParser.ArrayContext ctx) {
        List<lclangParser.ExpressionContext> expressionContexts = ctx.expression();
        Expression[] expressions = new Expression[expressionContexts.size()];

        for (int i = 0, l = expressions.length; i < l; i++) {
            expressions[i] = visitExpression(expressionContexts.get(i));
        }

        return new ArrayExpression(expressions);
    }

    @Override
    public Statement visitReturnExpr(lclangParser.ReturnExprContext ctx) {
        Expression returnExpression;
        lclangParser.ExpressionContext expressionContext = ctx.expression();
        if(expressionContext!=null)
            returnExpression = visitExpression(expressionContext);
        else returnExpression = VariableExpression.get("void");

        return new ReturnExpression(returnExpression);
    }

    @Override
    public Statement visitTypeGet(lclangParser.TypeGetContext ctx) {
        int startLine = ctx.getStart().getLine();

        return new AccessExpression(
                visitExpression(ctx.expression()),
                VariableExpression.get(PostVMClass.Constants.FIELD_TYPE),
                startLine
        );
    }

    @Override
    public Statement visitNewClass(lclangParser.NewClassContext ctx) {
        return new GetClassConstructorExpression(
                ctx.className.getText(),
                ctx.getStart().getLine()
        );
    }

    @Override
    public Expression visitPrimitive(lclangParser.PrimitiveContext ctx) {
        return (Expression) ctx.children.get(0).accept(this);
    }

    @Override
    public Statement visitStop(lclangParser.StopContext ctx) {
        return StopExpression.STOP;
    }

    @Override
    public Expression visitExpression(lclangParser.ExpressionContext ctx) {
        lclangParser.PrimitiveContext primitiveContext = ctx.primitive();
        if(primitiveContext!=null) return (Expression) ctx.children.get(0).accept(this);

        List<lclangParser.ExpressionContext> expressionContexts = ctx.expression();
        List<Expression> expressions = new ArrayList<>();
        for (lclangParser.ExpressionContext expressionContext: expressionContexts)
            expressions.add(visitExpression(expressionContext));

        if(ctx.call!=null) {
            Expression callable = expressions.get(0);
            expressions.remove(0);

            return new CallExpression(
                    callable,
                    expressions.toArray(new Expression[0]),
                    ctx.getStart().getLine()
            );
        }else if(ctx.access!=null) {
            return new AccessExpression(
                    expressions.get(0),
                    visitVariable(ctx.variable()),
                    ctx.getStart().getLine()
            );
        }else if(ctx.is!=null) {
            return new IsExpression(
                    expressions.get(0),
                    getType(ctx.type())
            );
        }else if(expressions.size()==1){
            int unaryOperation = 0;
            int binOperation = 0;

            if(ctx.unaryPlus!=null)
                binOperation = BinaryOperationExpression.Operation.ADD;
            else if(ctx.unaryMinus!=null)
                binOperation = BinaryOperationExpression.Operation.MINUS;
            else if(ctx.not!=null)
                unaryOperation = UnaryOperationExpression.Operation.NOT;
            else if(ctx.throwNull!=null)
                unaryOperation = UnaryOperationExpression.Operation.NULL_CHECK;
            else if(ctx.arrayAccess!=null)
                unaryOperation = UnaryOperationExpression.Operation.ARRAY_ACCESS;
            else if(ctx.compliment!=null)
                unaryOperation = UnaryOperationExpression.Operation.BINARY_COMPLIMENT;

            if(unaryOperation!=0) return new UnaryOperationExpression(
                    expressions.get(0),
                    unaryOperation,
                    ctx.getStart().getLine()
            );

            return new AssignExpression(expressions.get(0), new BinaryOperationExpression(
                    expressions.get(0),
                    ValueExpression.get(1),
                    binOperation,
                    ctx.getStart().getLine()
            ));
        }

        int operation = 0;
        if(ctx.add!=null)
            operation = BinaryOperationExpression.Operation.ADD;
        else if(ctx.minus!=null)
            operation = BinaryOperationExpression.Operation.MINUS;
        else if(ctx.multiplication!=null)
            operation = BinaryOperationExpression.Operation.MULTIPLICATION;
        else if(ctx.div!=null)
            operation = BinaryOperationExpression.Operation.DIVISION;
        else if(ctx.pow!=null)
            operation = BinaryOperationExpression.Operation.POW;
        else if(ctx.largerEquals!=null)
            operation = BinaryOperationExpression.Operation.LARGER_EQUALS;
        else if(ctx.larger!=null)
            operation = BinaryOperationExpression.Operation.LARGER;
        else if(ctx.lessEquals!=null)
            operation = BinaryOperationExpression.Operation.LESS_EQUALS;
        else if(ctx.less!=null)
            operation = BinaryOperationExpression.Operation.LESS;
        else if(ctx.equals!=null)
            operation = BinaryOperationExpression.Operation.EQUALS;
        else if(ctx.notEquals!=null)
            operation = BinaryOperationExpression.Operation.NOT_EQUALS;
        else if(ctx.or!=null)
            operation = BinaryOperationExpression.Operation.OR;
        else if(ctx.andBin!=null)
            operation = BinaryOperationExpression.Operation.BINARY_AND;
        else if(ctx.xor!=null)
            operation = BinaryOperationExpression.Operation.BINARY_XOR;
        else if(ctx.orBin!=null)
            operation = BinaryOperationExpression.Operation.BINARY_OR;
        else if(ctx.leftShift!=null)
            operation = BinaryOperationExpression.Operation.BINARY_LEFT_SHIFT;
        else if(ctx.rightShift!=null)
            operation = BinaryOperationExpression.Operation.BINARY_RIGHT_SHIFT;
        else if(ctx.and!=null)
            operation = BinaryOperationExpression.Operation.AND;
        else if(ctx.arrayAccess!=null)
            operation = BinaryOperationExpression.Operation.ARRAY_ACCESS;
        else if(ctx.nullableOr!=null)
            operation = BinaryOperationExpression.Operation.NULLABLE_OR;

        if(ctx.assign!=null) {
            Expression setExpression = expressions.get(1);
            if(operation!=0)
                setExpression = new BinaryOperationExpression(
                        expressions.get(0),
                        setExpression, operation,
                        ctx.getStart().getLine());

            return new AssignExpression(
                    expressions.get(0),
                    setExpression
            );
        }

        return new BinaryOperationExpression(
                expressions.get(0),
                expressions.get(1),
                operation,
                ctx.getStart().getLine()
        );
    }

    @Override
    public Expression visitParentnesesExpr(lclangParser.ParentnesesExprContext ctx) {
        return visitExpression(ctx.expression());
    }

    @Override
    public Statement visitLambda(lclangParser.LambdaContext ctx) {
        lclangParser.TypeContext returnTypeCtx = ctx.type();
        TypeStatement returnType = NamedTypeStatement.VOID;
        if(returnTypeCtx!=null)
            returnType = getType(returnTypeCtx);

        return new LambdaExpression(
                getArguments(ctx.arg()),
                returnType,
                visitExpression(ctx.expression()),
                ctx.getStart().getLine()
        );
    }

    public MethodStatement.Argument[] getArguments(List<lclangParser.ArgContext> argsCtx) {
        MethodStatement.Argument[] args = new MethodStatement.Argument[argsCtx.size()];
        for (int i = 0, l = args.length; i < l; i++) {
            lclangParser.ArgContext argCtx = argsCtx.get(i);

            TypeStatement type = NamedTypeStatement.ANY;
            lclangParser.TypeContext typeContext = argCtx.type();
            if(typeContext!=null)
                type = getType(typeContext);

            args[i] = new MethodStatement.Argument(argCtx.ID().getText(), type);
        }

        return args;
    }

    public void fillFile(LCLangFileClass root, lclangParser.FileContext ctx) throws LCLangRuntimeException {
        root.parents.addAll(Global.javaLibraries);
        root.parents.addAll(Global.libraries);

        List<lclangParser.GlobalContext> globalsCtx = ctx.global();
        for(lclangParser.GlobalContext globalContext: globalsCtx)
            root.globals.put(
                    globalContext.ID().getText(),
                    visitValue(globalContext.value()).visit(null, root.executor).get()
            );

        List<Statement> statements = new ArrayList<>();
        List<lclangParser.StmtContext> statementsCtx = ctx.stmt();
        for(lclangParser.StmtContext stmtContext: statementsCtx)
            statements.add(visitStmt(stmtContext));

        root.globals.put(LCLangFileClass.MAIN_METHOD_NAME, new MethodImpl(root.executor, new MethodStatement.Argument[]{
                new MethodStatement.Argument("args", new NamedTypeStatement("array", 0))
        }, VoidClass.PROTOTYPE.type, new BlockStatement(statements.toArray(new Statement[0])), false));

        List<lclangParser.ClassExprContext> classExprContexts = ctx.classExpr();
        for(lclangParser.ClassExprContext classExprContext: classExprContexts)
            getClass(classExprContext).visit(root);

        List<lclangParser.MethodContext> methodContexts = ctx.method();
        for(lclangParser.MethodContext methodContext: methodContexts) {
            MethodStatement method = getMethod(methodContext);
            root.globals.put(method.name, method.visit(root, false));
        }

        System.gc();
    }

    public ClassStatement getClass(lclangParser.ClassExprContext ctx) {
        List<Statement> statements = new ArrayList<>();
        List<lclangParser.StmtContext> stmtContexts = ctx.stmt();
        for (lclangParser.StmtContext stmtContext: stmtContexts) {
            statements.add(visitStmt(stmtContext));
        }

        List<lclangParser.MethodContext> methodContexts = ctx.method();
        MethodStatement[] methodStatements = new MethodStatement[methodContexts.size()];
        for (int i = 0; i < methodStatements.length; i++) {
            methodStatements[i] = getMethod(methodContexts.get(i));
        }

        NamedTypeStatement extendsClass = null;
        Expression[] extendsArguments = new Expression[0];
        if(ctx.classExtends!=null) {
            extendsClass = getNamedType(ctx.classExtends);

            lclangParser.ExtendsArgsContext extendsArgCtx = ctx.extendsArgs();
            if(extendsArgCtx!=null) {

                List<lclangParser.ExpressionContext> expressionContexts = extendsArgCtx.expression();
                extendsArguments = new Expression[expressionContexts.size()];

                for (int i = 0, l = extendsArguments.length; i < l; i++) {
                    extendsArguments[i] = visitExpression(expressionContexts.get(i));
                }
            }
        }

        return new ClassStatement(
                ctx.name.getText(),
                extendsClass,
                getArguments(ctx.arg()),
                statements,
                extendsArguments,
                methodStatements
        );
    }

    public MethodStatement getMethod(lclangParser.MethodContext ctx) {
        lclangParser.TypeContext returnTypeCtx = ctx.type();
        TypeStatement returnType = NamedTypeStatement.VOID;
        if(returnTypeCtx!=null) returnType = getType(returnTypeCtx);

        Statement statement;
        lclangParser.ExpressionContext expressionContext = ctx.expression();
        if(expressionContext!=null)
            statement = visitExpression(expressionContext);
        else statement = visitBlock(ctx.block());

        return new MethodStatement(
                ctx.ID().getText(),
                getArguments(ctx.arg()),
                returnType, statement);
    }

    public TypeStatement getType(lclangParser.TypeContext ctx) {
        lclangParser.PrimaryTypeContext primaryTypeContext = ctx.primaryType();
        if(primaryTypeContext!=null) return getPrimaryType(primaryTypeContext);

        List<lclangParser.PrimaryTypeContext> unionTypeContext = ctx.unionType().primaryType();
        TypeStatement[] typeStatements = new TypeStatement[unionTypeContext.size()];
        for (int i = 0, l = typeStatements.length; i < l; i++) {
            typeStatements[i] = getPrimaryType(unionTypeContext.get(i));
        }

        return new UnionTypeStatement(typeStatements);
    }

    public TypeStatement getPrimaryType(lclangParser.PrimaryTypeContext ctx) {
        lclangParser.MethodTypeContext typeContext = ctx.methodType();

        TypeStatement type;
        if(typeContext!=null)
            type = getMethodType(typeContext);
        else type = getNamedType(ctx.namedType());

        type.nullable = ctx.nullable!=null;
        return type;
    }

    private NamedTypeStatement getNamedType(lclangParser.NamedTypeContext namedTypeContext) {
        return new NamedTypeStatement(
                namedTypeContext.getText(),
                namedTypeContext.getStart().getLine()
        );
    }

    private CallableTypeStatement getMethodType(lclangParser.MethodTypeContext typeContext) {
        List<lclangParser.TypeContext> typeCtxList = typeContext.type();
        TypeStatement returnType = getType(typeCtxList.get(typeCtxList.size()-1));
        typeCtxList.remove(typeCtxList.size()-1);

        List<TypeStatement> typeStatements = new ArrayList<>();
        for (lclangParser.TypeContext type: typeCtxList) {
            typeStatements.add(getType(type));
        }

        return new CallableTypeStatement(typeStatements.toArray(new TypeStatement[0]),
                returnType,
                typeContext.getStart().getLine()
        );
    }
}

File: lclang/src/main/kotlin/lclang/LCCompiler.kt
package lclang

import lclang.exceptions.ClassNotFoundException
import lclang.libs.lang.*
import lclang.methods.MethodImpl
import lclang.statements.*
import lclang.statements.expressions.*
import lclang.types.Types

class LCCompiler: lclangBaseVisitor<Statement>() {
    override fun visitStmt(ctx: lclangParser.StmtContext): Statement = visit(ctx.children[0])
    override fun visitVariable(ctx: lclangParser.VariableContext) =
        VariableExpression(
            ctx.ID().text,
            ctx.start.line,
            ctx.stop.line
        )
    override fun visitBlock(ctx: lclangParser.BlockContext): Statement =
        BlockStatement(ctx.stmt().map { visitStmt(it) }.toTypedArray(), ctx.start.line, ctx.stop.line)
    override fun visitContainer(ctx: lclangParser.ContainerContext): Expression =
        ContainerExpression(ctx.stmt().map { visitStmt(it) }
            .toTypedArray(), ctx.start.line, ctx.stop.line)

    override fun visitValue(ctx: lclangParser.ValueContext): ValueExpression {
        val text = ctx.text
        return ValueExpression(
            when {
                ctx.STRING() != null -> StringClass(
                    unescapeString(text.substring(1).substringBeforeLast('"'))
                ).asValue()
                ctx.CHAR() != null -> CharClass(unescapeString(text)[1]).asValue()
                ctx.DOUBLE() != null -> DoubleClass(text.toDouble()).asValue()
                ctx.INTEGER() != null -> IntClass(text.toInt()).asValue()
                ctx.LONG() != null -> LongClass(text.substringBeforeLast('L').toLong()).asValue()
                ctx.HEX_LONG() != null -> LongClass(text.substring(1).toLong(radix = 16)).asValue()
                ctx.BOOL() != null -> BoolClass(text.startsWith('t')).asValue()
                ctx.NULL() != null -> Value(Types.UNDEFINED, null as LCClass?, Value.State.NOTHING)

                else -> throw Exception("Invalid value $text")
//            else -> throw LCLangException("Syntax error", "Invalid value $text", ctx.asCaller(this))
            }, ctx.start.line, ctx.stop.line
        )
    }

    override fun visitClassExpr(ctx: lclangParser.ClassExprContext): Statement {
        return ClassStatement(ctx.name.text, ctx.classExtends.text, ctx.start.line, ctx.stop.line)
    }

//    override fun visitMethod(ctx: lclangParser.MethodContext): Statement {
//        return MethodStatement(ctx.ID().text, ctx.args().arg().map {
//            it.ID().text to Types.getType(it.type())
//        }.toMap())
//    }

    override fun visitWhileStmt(ctx: lclangParser.WhileStmtContext): Statement {
        return WhileStatement(visitExpression(ctx.expression()), ctx.stmt()?.let { visitStmt(it) },
            ctx.start.line, ctx.stop.line)
    }

    override fun visitIfStmt(ctx: lclangParser.IfStmtContext): Statement = IfStatement(
        visitExpression(ctx.condition),
        visitStmt(ctx.ifT),
        ctx.ifF?.let { visitStmt(it) },
        ctx.start.line, ctx.stop.line)

    override fun visitIfExpr(ctx: lclangParser.IfExprContext): Expression =
        IfExpression(
            visitExpression(ctx.condition),
            visitExpression(ctx.ifT),
            visitExpression(ctx.ifF),
            ctx.start.line, ctx.stop.line
        )

    override fun visitArray(ctx: lclangParser.ArrayContext): Expression =
        ArrayExpression(
            ctx.expression().map { visitExpression(it) }, ctx.start.line, ctx.stop.line
        )

    override fun visitReturnExpr(ctx: lclangParser.ReturnExprContext): Expression =
        ReturnExpression(
            ctx.expression()?.let { visitExpression(it) },
            ctx.start.line,
            ctx.stop.line
        )

    override fun visitTypeGet(ctx: lclangParser.TypeGetContext): Expression =
        GetTypeExpression(
            visitExpression(ctx.expression()),
            ctx.start.line,
            ctx.stop.line
        )

    override fun visitNewClass(ctx: lclangParser.NewClassContext): Expression =
        GetClassConstructorExpression(
            ctx.className.text,
            ctx.start.line,
            ctx.stop.line
        )


    fun visitType_(ctx: lclangParser.TypeContext): TypeStatement {
        return when {
            ctx.methodType() != null -> visitCallableType(ctx.methodType())
            else -> visitNamedType_(ctx.namedType())
        }
    }

    override fun visitLambda(ctx: lclangParser.LambdaContext): Statement {
        return LambdaExpression(ctx.arg().associate {
            it.ID().text to (it.type()?.let { type -> visitType_(type)})
        }, ctx.type()?.let { visitType_(it) },
            visitExpression(ctx.expression()), ctx.start.line, ctx.stop.line)
    }

    fun visitNamedType_(ctx: lclangParser.NamedTypeContext): NamedTypeStatement {
        return NamedTypeStatement(ctx.text)
    }

    fun visitCallableType(ctx: lclangParser.MethodTypeContext): CallableTypeStatement {
        val args = ctx.type()
        val returnType = ctx.type().removeLast()
        return CallableTypeStatement(args.map {
            visitType_(it)
        }.toTypedArray(), visitType_(returnType))
    }

    override fun visitExpression(ctx: lclangParser.ExpressionContext): Expression =
        ctx.primitive()?.let { visitPrimitive(it) } ?:
        kotlin.run {
            val expressions = ctx.expression().map { visitExpression(it) }
            if(ctx.call!=null){
                val arguments = expressions.toMutableList()
                arguments.removeAt(0)

                CallExpression(
                    expressions[0],
                    arguments,
                    ctx.call.line,
                    ctx.call.line
                )
            }else if(ctx.access!=null){
                AccessExpression(
                    expressions[0],
                    visitVariable(ctx.variable()),
                    ctx.start.line,
                    ctx.stop.line
                )
            }else if(expressions.size==1){
                UnaryOperationExpression(expressions[0],
                    when {
                        ctx.unaryPlus!=null -> UnaryOperationExpression.Operation.UNARY_PLUS
                        ctx.unaryMinus!=null -> UnaryOperationExpression.Operation.UNARY_MINUS
                        ctx.not!=null -> UnaryOperationExpression.Operation.NOT
                        ctx.throwNull!=null -> UnaryOperationExpression.Operation.NULL_CHECK
                        else -> throw Exception("fe "+ctx.text)
                    }, ctx.start.line, ctx.stop.line)
            }else{
                val binaryOperation = when {
                    ctx.`is`!=null -> BinaryOperationExpression.Operation.IS
                    ctx.add!=null -> BinaryOperationExpression.Operation.ADD
                    ctx.minus!=null -> BinaryOperationExpression.Operation.MINUS
                    ctx.multiplication!=null -> BinaryOperationExpression.Operation.MULTIPLICATION
                    ctx.div!=null -> BinaryOperationExpression.Operation.DIVISION
                    ctx.pow!=null -> BinaryOperationExpression.Operation.POW
                    ctx.largerEquals!=null -> BinaryOperationExpression.Operation.LARGER_EQUALS
                    ctx.larger!=null -> BinaryOperationExpression.Operation.LARGER
                    ctx.less!=null -> BinaryOperationExpression.Operation.LESS
                    ctx.lessEquals!=null -> BinaryOperationExpression.Operation.LESS_EQUALS
                    ctx.equals!=null -> BinaryOperationExpression.Operation.EQUALS
                    ctx.notEquals!=null -> BinaryOperationExpression.Operation.NOT_EQUALS
                    ctx.or!=null -> BinaryOperationExpression.Operation.OR
                    ctx.and!=null -> BinaryOperationExpression.Operation.AND
                    ctx.arrayAccess!=null -> BinaryOperationExpression.Operation.ARRAY_ACCESS
                    ctx.nullableOr!=null -> BinaryOperationExpression.Operation.NULLABLE_OR
                    else -> null
                }

                if(ctx.assign!=null) {
                    var setExpression = expressions[1]
                    if(binaryOperation!=null)
                        setExpression = BinaryOperationExpression(expressions[0],
                            setExpression, binaryOperation, ctx.start.line, ctx.stop.line)

                    AssignExpression(
                        expressions[0],
                        setExpression,
                        ctx.start.line,
                        ctx.stop.line
                    )
                } else BinaryOperationExpression(expressions[0], expressions[1], binaryOperation ?:
                        throw Exception("fe"), ctx.start.line, ctx.stop.line)
            }
        }

    fun fillFile(root: LCRootExecutor, ctx: lclangParser.FileContext) {
        val libraries = Global.libraries+Global.javaLibraries
        for(library in libraries) root.accept(library)

        root.statements.clear()
        root.statements.addAll(ctx.stmt().map { visitStmt(it) })
        root.globals.putAll(ctx.global().associate {
            it.ID().text to visitValue(it.value()).value
        })

        val caller = Caller(root, 0, 0)
        root.classes.putAll(ctx.classExpr().associate { clazzAst ->
            clazzAst.name.text to object: LCClass(clazzAst.name.text, root.path) {
                init {
                    constructor = constructor {
                        val clazz = LCClass(name, prevCaller.root.path)
                        if(clazzAst.classExtends!=null) {
                            clazz.extendsClass = prevCaller.root.classes[clazzAst.classExtends.text]
                                ?: throw ClassNotFoundException(
                                    name,
                                    prevCaller
                                )

                            clazz.accept(clazz.extendsClass!!)
                        }

                        clazz.executor.variables.putAll(executor.variables)
                        clazz.accept(prevCaller.root)

                        for(method in clazzAst.method())
                            clazz.globals[method.ID().text] = MethodImpl(clazz.executor, method.arg().associate {
                                it.ID().text to (it.type()?.let { type -> visitType_(type).toType(caller)}
                                    ?: Types.ANY)
                            }, method.returnType?.let { type -> visitType_(type).toType(caller)}
                                ?: Types.VOID,
                                visit(method.block()?:method.expression())).asValue()

                        for (stmt in clazzAst.stmt())
                            visitStmt(stmt).visit(Caller(clazz, 0, 0), clazz.executor)

                        return@constructor clazz
                    }
                }
            }
        })

        root.globals.putAll(ctx.method().associate { method ->
            method.ID().text to MethodImpl(root.executor, method.arg().associate {
                it.ID().text to (it.type()?.let { type -> visitType_(type).toType(caller)}
                    ?: Types.ANY)
            }, method.returnType?.let { type -> visitType_(type).toType(caller)}
                ?: Types.VOID,
                visit(method.block()?:method.expression())).asValue()
        })
    }


    override fun visitStop(ctx: lclangParser.StopContext): Expression =
        StopExpression(ctx.start.line, ctx.stop.line)
    override fun visitPrimitive(ctx: lclangParser.PrimitiveContext): Expression = visit(ctx.children[0]) as Expression
}